src/java.base/share/classes/java/util/ReverseOrderListView.java

```diff
@@ -0,0 +1,381 @@
+     * method. For bulk ops (addAll, removeAll, etc.) this throws unconditionally.
+     * In contrast, if the base list inherits a bulk op implementation from AbstractList,
+     * it might not throw if no actual mutation would be attempted (e.g., addAll on an
+     * empty collection). Arguably calling this is unnecessary for individual ops,
+     * for which the base list should always throw, but it's easier to verify the right
+     * behavior if every mutator of this class always checks.
+     */
+    void checkModifiable() {
+        if (! modifiable)
+            throw new UnsupportedOperationException();
+    }
+
+    class DescendingIterator implements Iterator<E> {
+        final ListIterator<E> it = base.listIterator(base.size());
+        public boolean hasNext() { return it.hasPrevious(); }
+        public E next() { return it.previous(); }
+        public void remove() {
+            checkModifiable();
+            it.remove();
+            // TODO - make sure ListIterator is positioned correctly afterward
+        }
+    }
+
+    class DescendingListIterator implements ListIterator<E> {
+        final ListIterator<E> it;
+
+        DescendingListIterator(int size, int pos) {
+            if (pos < 0 || pos > size)
+                throw new IndexOutOfBoundsException();
+            it = base.listIterator(size - pos);
+        }
+
+        public boolean hasNext() {
+            return it.hasPrevious();
+        }
+
+        public E next() {
+            return it.previous();
+        }
+
+        public boolean hasPrevious() {
+            return it.hasNext();
+        }
+
+        public E previous() {
+            return it.next();
+        }
+
+        public int nextIndex() {
+            return base.size() - it.nextIndex();
+        }
+
+        public int previousIndex() {
+            return nextIndex() - 1;
+        }
+
+        public void remove() {
+            checkModifiable();
+            it.remove();
+        }
+
+        public void set(E e) {
+            checkModifiable();
+            it.set(e);
+        }
+
+        public void add(E e) {
+            checkModifiable();
+            it.add(e);
+            it.previous();
+        }
+    }
+
+    // ========== Iterable ==========
+
+    public void forEach(Consumer<? super E> action) {
+        for (E e : this)
+            action.accept(e);
+    }
+
+    public Iterator<E> iterator() {
+        return new DescendingIterator();
+    }
+
+    public Spliterator<E> spliterator() {
+        // TODO can probably improve this
+        return Spliterators.spliteratorUnknownSize(new DescendingIterator(), 0);
+    }
+
+    // ========== Collection ==========
+
+    public boolean add(E e) {
+        checkModifiable();
+        base.add(0, e);
+        return true;
+    }
+
+    public boolean addAll(Collection<? extends E> c) {
+        checkModifiable();
+        boolean modified = false;
+        for (E e : c) {
+            base.add(0, e);
```

Comment:
This is worrysome from performance point of view for base implementations like ArrayList where every insertion means shifting the whole array. At least, we could optimize if `c` is `SequencedCollection`, like `if (c instanceof SequencedCollection<? extends E> sc) base.addAll(0, sc.reversed())`. Otherwise, we can use `base.addAll(0, Arrays.asList(sc.toArray()).reversed())` (with a couple of unchecked casts).

Reply:
Fixed.